eta/core/serial.py

```diff
@@ -1124,17 +1113,477 @@ def _filter_elements(self, filters, match):
+                "Cannot use %s to reflectively load this container "
+                "because the expected field '%s' was not found in the "
+                "JSON dictionary" % (cls, cls._ELE_CLS_FIELD))
+
+        return etau.get_class(d["_CLS"])
+
+
+class BigContainer(Container):
+    '''Container that stores a (potentially huge) list of `Serializable`
+    objects. The elements are stored on disk in a backing directory; accessing
+    any element in the list causes an immediate READ from disk, and
+    adding/setting an element causes an immediate WRITE to disk.
+
+    BigContainers store a `backing_dir` attribute that specifies the path on
+    disk to the serialized elements. The container also maintains a list of
+    uuids in its `_ELE_ATTR` field to locate elements on disk. This list is
+    included in lieu of the actual elements when serializing BigContainer
+    instances.
+
+    To read/write archives of BigContainer that also include their elements,
+    use the `to_archive()` and `from_archive()` methods, respectively.
+
+    This class cannot be instantiated directly. Instead a subclass should
+    be created for each type of element to be stored. Subclasses MUST set the
+    following members:
+        -  `_ELE_CLS`: the class of the element stored in the container
+
+    In addition, sublasses MAY override the following members:
+        - `_ELE_CLS_FIELD`: the name of the private attribute that will store
+            the class of the elements in the container
+        - `_ELE_ATTR`: the name of the attribute that will store the elements
+            in the container
+
+    Examples:
+        ```
+        import eta.core.geometry as etag
+
+        point = etag.LabeledPoint("origin", etag.RelativePoint(0, 0))
+        points = etag.BigLabeledPointContainer("/tmp/BigLabeledPointContainer")
+
+        # Immediately writes the LabeledPoint to disk
+        points.add(point)
+
+        # Reads the LabeledPoint from disk
+        print(points[0])
+
+        # Only the index of the BigContainer is serialized
+        print(points)
+        ```
+    '''
+
+    def __init__(self, backing_dir, **kwargs):
+        '''Creates a BigContainer instance.
+
+        Args:
+            backing_dir: the backing directory in which the elements are/will
+                be stored
+            <elements>: an optional list of uuids for elements in the
+                container. The appropriate name of this keyword argument is
+                determined by the `_ELE_ATTR` member of the BigContainer
+                subclass
+
+        Raises:
+            ContainerError: if there was an error while creating the container
+        '''
+        self._validate_cls()
+
+        self._backing_dir = None
+        self._set_backing_dir(backing_dir)
+
+        if self._ELE_ATTR in kwargs:
+            etau.ensure_dir(self.backing_dir)
+            elements = kwargs[self._ELE_ATTR]
+        else:
+            etau.ensure_empty_dir(self.backing_dir)
```

Comment:
Note that I removed the code that would try to automatically populate a BigContainer based on the contents of the backing directory. I did this because I think `BigContainer(backing_dir)` should throw an error if `backing_dir` is not empty.

The user can use `BigContainer.from_paths` or `BigContainer.from_dir` if they want to automatically populate a `BigContainer` from disk